datafusion/core/src/physical_plan/unnest.rs

```diff
@@ -250,126 +249,266 @@ fn build_batch(
+                }
+                take_offset += array.len();
+            }
+            None => {
+                take_indicies_builder.append_null();
+            }
+        });
+        Ok(kernels::take::take(
+            &values,
+            &take_indicies_builder.finish(),
+            None,
+        )?)
+    }
+}
+
+fn build_batch_fixedsize_list(
+    batch: &RecordBatch,
+    schema: &SchemaRef,
+    unnest_column_idx: usize,
+    list_array: &FixedSizeListArray,
+    options: &UnnestOptions,
+) -> Result<RecordBatch> {
+    let unnested_array = unnest_fixed_list(list_array, options)?;
+
+    let take_indicies =
+        create_take_indicies_fixed(list_array, unnested_array.len(), options);
+
+    batch_from_indices(
+        batch,
+        schema,
+        unnest_column_idx,
+        &unnested_array,
+        &take_indicies,
+    )
+}
+
+/// Given this `FixedSizeListArray` list_array:
+///   
+/// ```ignore
+/// [1, 2], null, [3, 4], null, [5, 6]
+/// ```
+/// Its values array is represented like this:
 ///
 /// ```ignore
-/// 1, null,  3, null, 2
+/// [1, 2, null, null 3, 4, null, null, 5, 6]
 /// ```
 ///
-/// the result indices array is:
+/// So if there are no null values
+/// we can return the values array without any copying.
+///
+/// Otherwise we'll transfrom the values array using the take kernel.
+///
+/// If `UnnestOptions.preserve_nulls` is true the take indicies will look like this:
 ///
 /// ```ignore
-/// 0, 1, 2, 2, 2, 3, 4, 4
+/// 0, 1, null, 4, 5, null, 8, 9
+/// ```
+/// Otherwise we drop the nulls and take indicies will look like this:
+///
+/// ```ignore
+/// 0, 1, 4, 5, 8, 9
 /// ```
 ///
-/// where a null value count as one element.
-fn create_take_indices<T>(
-    list_lengths: &PrimitiveArray<T>,
+fn unnest_fixed_list(
+    list_array: &FixedSizeListArray,
+    options: &UnnestOptions,
+) -> Result<Arc<dyn Array + 'static>> {
+    let values = list_array.values();
+
+    if list_array.null_count() == 0 {
+        Ok(values.clone())
+    } else {
+        let len_without_nulls =
+            values.len() - list_array.null_count() * list_array.value_length() as usize;
+        let null_count = if options.preserve_nulls {
+            list_array.null_count()
+        } else {
+            0
+        };
+        let mut builder =
+            PrimitiveArray::<Int32Type>::builder(len_without_nulls + null_count);
+        let mut take_offset = 0;
+        let fixed_value_length = list_array.value_length() as usize;
+        list_array.iter().for_each(|elem| match elem {
+            Some(_) => {
+                for i in 0..fixed_value_length {
+                    //take_offset + i is always positive
+                    let take_index =
+                        <Int32Type as ArrowPrimitiveType>::Native::from_usize(
+                            take_offset + i,
+                        )
+                        .unwrap();
+                    builder.append_value(take_index);
```

Comment:
This might be easier to understand
```suggestion
                    let take_index = take_offset + i;
                    builder.append_value(take_index as i32);
```